apis/Google.Cloud.Diagnostics.AspNet/Google.Cloud.Diagnostics.AspNet/Trace/CloudTrace.cs

```diff
@@ -119,16 +117,11 @@ public static void Initialize(string projectId, HttpApplication application, Tra
         private void BeginRequest(object sender, EventArgs e)
         {
             TraceHeaderContext headerContext = TraceHeaderContextUtils.CreateContext(HttpContext.Current.Request);
-            TraceOptions headerOptions = _headerFactory.CreateOptions(headerContext);
-
             // If the trace header says to trace or if the rate limiter allows tracing continue.
-            if (!headerOptions.ShouldTrace)
+            if (headerContext.ShouldTrace == false || (headerContext.ShouldTrace == null &&
+                !_rateFactory.CreateOptions().ShouldTrace))
```

Comment:
This condition takes a lot of mental gymnastics to read (and could therefore be a potential source of bugs with future maintenance), so I've create a table to verify my understanding:

|    | RF ShouldTrace = true | RF ShouldTrace = false  |
|---|---|---|
| HC ShouldTrace = true | trace | trace |
| HC ShouldTrace = false |  |  |
| HC ShouldTrace = null | trace      |  |

Does this accurately represent the intent? If so, I would suggest making a helper property or method named something like `ShouldTraceResolved` which does something like this, so its very clear:
```csharp
if (headerContext.ShouldTrace != null)
{
  return headerContext.ShouldTrace.Value;
}
return _rateFactory.CreateOptions().ShouldTrace;
```

And then just make this condition `if (!ShouldTraceResolved)`